hook-common/src/pgqueue.rs

```diff
@@ -284,30 +315,56 @@ impl<J: std::marker::Send, M: std::marker::Send> PgQueueJob for PgJob<J, M> {
 #[derive(Debug)]
 pub struct PgTransactionJob<'c, J, M> {
     pub job: Job<J, M>,
-    pub transaction: sqlx::Transaction<'c, sqlx::postgres::Postgres>,
+
+    /// The open transaction this job came from. If multiple jobs were queried at once, then this
+    /// transaction will be shared between them (across async tasks and threads as necessary). See
+    /// below for more information.
+    shared_txn: Arc<Mutex<Option<sqlx::Transaction<'c, sqlx::postgres::Postgres>>>>,
+}
+
+// Container struct for a batch of PgTransactionJob. Includes a reference to the shared transaction
+// for committing the work when all of the jobs are finished.
+pub struct PgTransactionBatch<'c, J, M> {
+    pub jobs: Vec<PgTransactionJob<'c, J, M>>,
+
+    /// The open transaction the jobs in the Vec came from. This should be used to commit or
+    /// rollback when all of the work is finished.
+    shared_txn: Arc<Mutex<Option<sqlx::Transaction<'c, sqlx::postgres::Postgres>>>>,
+}
+
+impl<'c, J, M> PgTransactionBatch<'_, J, M> {
+    pub async fn commit(&mut self) -> PgQueueResult<()> {
+        let mut txn_guard = self.shared_txn.lock().await;
+
+        let txn = txn_guard.take().unwrap();
```

Comment:
What is this `take` call doing? From the `Mutex` [docs](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#examples), it seems like we should deref `txn_guard`.

Reply:
Commit takes ownership, so taking it from an `Option` is the only way I found to do this. I wouldn't say we're _replacing_ it but yes we are _taking_ it. Let me know if I'm missing something. :)

Reply:
Doesn't look like you are taking ownership, at least I see just a mutable reference to `PgTransactionBatch`. Why not take ownership instead? At least semantically it seems it should be fine to be done with the batch once we commit.

I tried this out locally to see if there would be any issues, and there are none (besides needing to bring the `sqlx::postgres::any::AnyConnectionBackend` trait into scope):

```rust
    pub async fn commit(self) -> PgQueueResult<()> {
        let mut txn_guard = self.shared_txn.lock().await;

        txn_guard
            .as_deref_mut()
            .ok_or(PgQueueError::TransactionAlreadyClosedError)?
            .commit()
            .await
            .map_err(|e| PgQueueError::QueryError {
                command: "COMMIT".to_owned(),
                error: e,
            })?;

        Ok(())
    }
```

Reply:
Not very important though, I understood your explanation, thanks!

Reply:
> I kind of prefer this as sqlx takes ownership of transaction on commit, so it kind of makes sense that we take ownership of the transaction _batch_ when we commit it.

Yeah, makes sense to me, thanks!